File: node/src/Automator.ts
import { BrowserManager } from './BrowserManager';
import { ChatGPTProvider } from './providers/ChatGPT';

export interface LoggerInterface {
    info(message: string): void;
    warn(message: string): void;
    error(message: string): void;
    debug(message: string): void;
}

export interface AutomatorConfig {
    selectors?: Record<string, string>;
    logger?: LoggerInterface;
    onOtpRequired?: () => Promise<string>;
}

export class Automator {
    private browserManager!: BrowserManager;
    private provider: ChatGPTProvider | null = null;
    private providerName: string;
    private headless: boolean;
    private credentials?: { email?: string, password?: string, method?: string };
    private sessionPath?: string;
    private config: AutomatorConfig;
    private logger: LoggerInterface;

    constructor(
        provider: string = "chatgpt",
        headless: boolean = true,
        credentials?: { email?: string, password?: string, method?: string },
        sessionPath?: string,
        config?: AutomatorConfig
    ) {
        this.providerName = provider;
        this.headless = headless;
        this.credentials = credentials;
        this.sessionPath = sessionPath;
        this.config = config || {};

        // Default logger to console if not provided
        this.logger = this.config.logger || console;
    }

    public async init(): Promise<void> {
        this.browserManager = new BrowserManager();
        // Pass sessionPath to start if we want to load cookies (though persistent context handles it, 
        // we might want to explicitly load a state file if provided)
        const page = await this.browserManager.start(this.headless, this.sessionPath);

        if (this.browserManager.context) {
            await this.browserManager.context.grantPermissions(["clipboard-read", "clipboard-write"]);
        }

        if (this.providerName === "chatgpt") {
            this.provider = new ChatGPTProvider(page, this.config.selectors, this.logger, this.config.onOtpRequired);
        } else {
            throw new Error(`Unknown provider: ${this.providerName}`);
        }

        // Check Auth / Auto-Login
        if (!await this.browserManager.isAuthenticated(this.provider.URL, this.provider.SEL_PROFILE_BTN)) {
            this.logger.info("Not authenticated. Initiating login...");

            // Credentials priority: Constructor args > Env vars
            const email = this.credentials?.email || process.env.CHATGPT_EMAIL;
            const password = this.credentials?.password || process.env.CHATGPT_PASSWORD;
            const method = this.credentials?.method || "email";

            if (!email || !password) {
                throw new Error("Credentials not provided. Set CHATGPT_EMAIL and CHATGPT_PASSWORD env vars or pass them to constructor.");
            }

            await this.provider.login({ email, password, method });

            if (this.sessionPath) {
                await this.browserManager.saveSession(this.sessionPath);
                this.logger.info(`Session saved to ${this.sessionPath}`);
            }
        } else {
            this.logger.info("Session authenticated.");
        }
    }

    public async processPrompt(prompt: string): Promise<string> {
        if (!this.provider) throw new Error("Provider not initialized. Call init() first.");
        return await this.provider.sendPrompt(prompt);
    }

    public async processChain(prompts: (string | ((prev: string) => string))[]): Promise<string[]> {
        const responses: string[] = [];
        let lastResponse = "";

        for (let i = 0; i < prompts.length; i++) {
            let currentPrompt = "";
            const promptItem = prompts[i];

            if (typeof promptItem === 'function') {
                currentPrompt = promptItem(lastResponse);
            } else {
                currentPrompt = promptItem;
                if (currentPrompt.includes("{}")) {
                    currentPrompt = currentPrompt.replace("{}", lastResponse);
                } else if (currentPrompt.includes("{{}}")) {
                    currentPrompt = currentPrompt.replace("{{}}", lastResponse);
                }
            }

            this.logger.info(`Processing prompt ${i + 1}/${prompts.length}...`);
            const response = await this.processPrompt(currentPrompt);
            responses.push(response);
            lastResponse = response;
        }

        return responses;
    }

    public async close(): Promise<void> {
        if (this.browserManager) {
            await this.browserManager.stop();
        }
    }
}

File: python/src/llm_session/automator.py
from typing import List, Optional, Callable, Union
import logging
from .browser import BrowserManager
from .config import Config
from .providers.chatgpt import ChatGPTProvider
from .exceptions import SetupError, OTPRequiredError

logger = logging.getLogger(__name__)

class Automator:
    """
    Main entry point for the LLM Web Automator.
    """

    def __init__(self, provider: str = "chatgpt", headless: bool = True, credentials: Optional[dict] = None, session_path: Optional[str] = None, config: Optional[dict] = None, on_otp_required: Optional[Callable[[], str]] = None):
        self.provider_name = provider
        self.headless = headless
        self.credentials = credentials
        self.session_path = session_path
        self.config = config or {}
        self.on_otp_required = on_otp_required
        self.browser_manager = BrowserManager()
        self.provider = None
        
        # Initialize
        self._setup()

    def _setup(self):
        """Initialize browser and provider."""
        # 1. Start Browser
        # Pass session_path if available to load cookies
        page = self.browser_manager.start(headless=self.headless, session_path=self.session_path)
        
        # Grant clipboard permissions for response extraction
        self.browser_manager.context.grant_permissions(["clipboard-read", "clipboard-write"])
        
        # 2. Initialize Provider
        if self.provider_name.lower() == "chatgpt":
            # Pass config for selectors if needed
            self.provider = ChatGPTProvider(page, config=self.config, on_otp_required=self.on_otp_required)
        else:
            raise NotImplementedError(f"Provider {self.provider_name} not supported.")
            
        # 3. Check Auth / Auto-Login
        if not self.browser_manager.is_authenticated(self.provider.URL, self.provider.SEL_PROFILE_BTN):
            logger.info("Not authenticated. Initiating login...")
            
            # Prioritize passed credentials, fallback to Config/Env vars
            creds = self.credentials or Config.get_credentials(self.provider_name)
            
            if not creds.get("email") or not creds.get("password"):
                raise SetupError("Credentials not found. Pass them to Automator() or set environment variables (CHATGPT_EMAIL, CHATGPT_PASSWORD).")
            
            self.provider.login(creds)
            
            # Save session if path provided
            if self.session_path:
                self.browser_manager.save_session(self.session_path)
                logger.info(f"Session saved to {self.session_path}")
        else:
            logger.info("Session authenticated.")

    def process_prompt(self, prompt: str) -> str:
        """
        Send a single prompt and get the response.
        """
        if not self.provider:
            raise SetupError("Provider not initialized.")
        return self.provider.send_prompt(prompt)

    def process_chain(self, prompts: List[Union[str, Callable[[str], str]]]) -> List[str]:
        """
        Process a chain of prompts. 
        Supports string formatting with {} (previous response injected) or callables.
        """
        responses = []
        last_response = ""
        
        for i, prompt_item in enumerate(prompts):
            current_prompt = ""
            
            if callable(prompt_item):
                current_prompt = prompt_item(last_response)
            elif isinstance(prompt_item, str):
                # If it's a string, try to format it with the last response if it contains placeholders
                # We use a specific placeholder {{previous}} to avoid accidental formatting of user code
                # But we also support {} for backward compatibility if it's simple
                
                if "{{previous}}" in prompt_item:
                     current_prompt = prompt_item.replace("{{previous}}", last_response)
                elif "{}" in prompt_item:
                    # Basic check to avoid formatting JSON or Python dicts
                    # If the string looks like code, we might skip this or warn?
                    # For now, we'll just try format, but this is risky as noted in review.
                    # Better approach: Only format if it's clearly a placeholder.
                    try:
                        current_prompt = prompt_item.format(last_response)
                    except ValueError:
                        # If format fails (e.g. mismatched braces), treat as literal
                        current_prompt = prompt_item
                elif "{{}}" in prompt_item: # Support legacy double braces
                    current_prompt = prompt_item.replace("{{}}", last_response)
                else:
                    current_prompt = prompt_item
            
            logger.info(f"Processing prompt {i+1}/{len(prompts)}...")
            response = self.process_prompt(current_prompt)
            responses.append(response)
            last_response = response
            
        return responses

    def close(self):
        """Clean up resources."""
        self.browser_manager.stop()
